File: services/api/src/ProtoFast.Api/Services/Screenplays/LibraryNames.cs
using System.Globalization;

namespace ProtoFast.Api.Services.Screenplays;

/// <summary>Ports the client's <c>library-names.ts</c>, wording included.</summary>
public static class LibraryNames
{
    public const int MaxLength = 255;

    /// <summary>Ignores case and surrounding space, as the client's <c>sameName</c> does.</summary>
    public static bool Same(string a, string b) =>
        string.Compare(a.Trim(), b.Trim(), CultureInfo.InvariantCulture, CompareOptions.IgnoreCase) == 0;

    /// <returns>The name trimmed.</returns>
    public static string Check(IEnumerable<(Guid Id, string Name)> entries, string? name, string noun, Guid? exceptId = null)
    {
        var trimmed = name?.Trim() ?? "";
        if (trimmed.Length == 0)
        {
            throw StoryErrors.Invalid($"A {noun} needs a name.");
        }

        if (trimmed.Length > MaxLength)
        {
            throw StoryErrors.Invalid($"Names are limited to {MaxLength} characters.");
        }

        foreach (var entry in entries)
        {
            if (entry.Id != exceptId && Same(entry.Name, trimmed))
            {
                throw StoryErrors.AlreadyExists($"There’s already a {noun} called “{entry.Name}”.");
            }
        }

        return trimmed;
    }

    public static int Hue(int hue) =>
        hue is >= 0 and <= 359 ? hue : throw StoryErrors.Invalid("Hues run from 0 to 359.");
}

File: clients/theplot/src/app/stories/library-names.ts
/** The API's column lengths. */
export const NAME_MAX = 255;
export const LABEL_MAX = 64;

/** Ignores case and surrounding space. */
export function sameName(a: string, b: string): boolean {
  return a.trim().localeCompare(b.trim(), undefined, { sensitivity: 'accent' }) === 0;
}

export function findByName<T extends { id: string; name: string }>(
  items: readonly T[],
  name: string,
  exceptId?: string,
): T | undefined {
  return items.find((item) => item.id !== exceptId && sameName(item.name, name));
}

export function nameProblem<T extends { id: string; name: string }>(
  items: readonly T[],
  name: string,
  noun: string,
  exceptId?: string,
): string | null {
  const trimmed = name.trim();
  if (!trimmed) {
    return `A ${noun} needs a name.`;
  }
  if (trimmed.length > NAME_MAX) {
    return `Names are limited to ${NAME_MAX} characters.`;
  }
  const clash = findByName(items, trimmed, exceptId);
  return clash ? `There’s already a ${noun} called “${clash.name}”.` : null;
}

export function labelProblem(labels: readonly string[], label: string): string | null {
  const trimmed = label.trim();
  if (!trimmed) {
    return 'Type a label first.';
  }
  if (trimmed.length > LABEL_MAX) {
    return `Labels are limited to ${LABEL_MAX} characters.`;
  }
  const clash = labels.find((l) => sameName(l, trimmed));
  return clash ? `“${clash}” is already on the list.` : null;
}
